libethereum/Executive.cpp

```diff
@@ -218,18 +218,23 @@ void Executive::initialize(Transaction const& _transaction)
 			m_excepted = TransactionException::InvalidNonce;
 			BOOST_THROW_EXCEPTION(InvalidNonce() << RequirementError((bigint)nonceReq, (bigint)m_t.nonce()));
 		}
-	}
 
-	// Avoid unaffordable transactions.
-	bigint gasCost = (bigint)m_t.gas() * m_t.gasPrice();
-	bigint totalCost = m_t.value() + gasCost;
-	if (m_s.balance(m_t.sender()) < totalCost)
+		// Avoid unaffordable transactions.
+		bigint gasCost = (bigint)m_t.gas() * m_t.gasPrice();
+		bigint totalCost = m_t.value() + gasCost;
+		if (m_s.balance(m_t.sender()) < totalCost)
+		{
+			clog(ExecutiveWarnChannel) << "Not enough cash: Require >" << totalCost << "=" << m_t.gas() << "*" << m_t.gasPrice() << "+" << m_t.value() << " Got" << m_s.balance(m_t.sender()) << "for sender: " << m_t.sender();
+			m_excepted = TransactionException::NotEnoughCash;
+			BOOST_THROW_EXCEPTION(NotEnoughCash() << RequirementError(totalCost, (bigint)m_s.balance(m_t.sender())) << errinfo_comment(m_t.sender().abridged()));
+		}
+		m_gasCost = (u256)gasCost;  // Convert back to 256-bit, safe now.
+	}
+	else if (m_t.gasPrice() != 0 || m_t.value() != 0 || m_t.nonce() != 0)
```

Comment:
I hope that this check won't be needed when we merge the changes in `Executive` from https://github.com/ethereum/cpp-ethereum/pull/3655
(then `verifyTransaction` will be called here, which has this check)